presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.unittest;
+
+import com.facebook.presto.connector.meta.RequiredFeatures;
+import com.facebook.presto.spi.ColumnMetadata;
+import com.facebook.presto.spi.ConnectorOutputTableHandle;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorTableHandle;
+import com.facebook.presto.spi.ConnectorTableMetadata;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.SchemaTablePrefix;
+import com.facebook.presto.spi.connector.Connector;
+import com.facebook.presto.spi.connector.ConnectorMetadata;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.facebook.presto.spi.transaction.IsolationLevel;
+import com.facebook.presto.testing.TestingConnectorSession;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Streams;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.facebook.presto.connector.meta.ConnectorFeature.CREATE_TABLE;
+import static com.facebook.presto.connector.meta.ConnectorFeature.DROP_TABLE;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public interface BaseMetadataTest
+        extends SPITest
+{
+    Map<String, Object> getTableProperties();
+
+    default List<ColumnMetadata> withSystemColumns(List<ColumnMetadata> expectedColumns)
+    {
+        return expectedColumns;
+    }
+
+    default List<String> systemSchemas()
+    {
+        return ImmutableList.of();
+    }
+
+    default SchemaTableName schemaTableName(String tableName)
+    {
+        return new SchemaTableName("default_schema", tableName);
+    }
+
+    default SchemaTableName schemaTableName(String schemaName, String tableName)
+    {
+        return new SchemaTableName(schemaName, tableName);
+    }
+
+    @Test
+    default void testEmptyMetadata()
+    {
+        ConnectorSession session = new TestingConnectorSession(ImmutableList.of());
+
+        run(ImmutableList.of(
```

Comment:
> Why do we need run method. Can't we just use:

Hive doesn't support some combinations of operations in the same transaction (adding a schema and listing schemas, IIRC).

> Still nicer looking approach would be to have withMetadata take multiple consumers:

https://github.com/prestodb/presto/pull/8672#discussion_r131481926